File: 0647-palindromic-substrings/0647-palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:

        count = 0
        for i in range(len(s)):
            count += self.getCount(i, i, s)
            count += self.getCount(i, i + 1, s)
        
        return count
    
    def getCount(self, start, end, s):
        count = 0
        while start >= 0 and end < len(s) and s[start] == s[end]:
            count += 1
            start -= 1
            end += 1
        
        return count




        

File: 0647-palindromic-substrings/0647-palindromic-substrings.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, 0));
        int res = 0;

        for (int l = 1; l <= n; l++) {
            int j = l - 1;
            for (int i = 0; i < n - l + 1; i++) {
                if (l == 1) dp[i][j] = true;
                else if (l == 2) dp[i][j] = s[i] == s[j];
                else dp[i][j] = s[i] == s[j] and dp[i + 1][j - 1];
                res += dp[i][j];
                j++;
            }
        }

        return res;
    }
};